core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -137,93 +150,91 @@ protected boolean shouldKeep(T row) {
     }
   }
 
-  private static class PositionSetDeleteFilter<T> extends Filter<T> {
-    private final Function<T, Long> rowToPosition;
-    private final PositionDeleteIndex deleteSet;
-
-    private PositionSetDeleteFilter(Function<T, Long> rowToPosition, PositionDeleteIndex deleteSet) {
-      this.rowToPosition = rowToPosition;
-      this.deleteSet = deleteSet;
-    }
-
-    @Override
-    protected boolean shouldKeep(T row) {
-      return !deleteSet.isDeleted(rowToPosition.apply(row));
-    }
-  }
-
-  private static class PositionStreamDeleteFilter<T> extends CloseableGroup implements CloseableIterable<T> {
+  private abstract static class PositionStreamDeleteIterable<T> extends CloseableGroup implements CloseableIterable<T> {
     private final CloseableIterable<T> rows;
-    private final Function<T, Long> extractPos;
-    private final CloseableIterable<Long> deletePositions;
+    private final CloseableIterator<Long> deletePosIterator;
+    private final Function<T, Long> rowToPosition;
+    private long nextDeletePos;
 
-    private PositionStreamDeleteFilter(CloseableIterable<T> rows, Function<T, Long> extractPos,
-                                       CloseableIterable<Long> deletePositions) {
+    PositionStreamDeleteIterable(CloseableIterable<T> rows, Function<T, Long> rowToPosition,
+                                 CloseableIterable<Long> deletePositions) {
       this.rows = rows;
-      this.extractPos = extractPos;
-      this.deletePositions = deletePositions;
+      this.rowToPosition = rowToPosition;
+      this.deletePosIterator = deletePositions.iterator();
     }
 
     @Override
     public CloseableIterator<T> iterator() {
-      CloseableIterator<Long> deletePosIterator = deletePositions.iterator();
-
       CloseableIterator<T> iter;
       if (deletePosIterator.hasNext()) {
-        iter = new PositionFilterIterator(rows.iterator(), deletePosIterator);
+        nextDeletePos = deletePosIterator.next();
+        iter = createPosDeleteIterator(rows.iterator());
       } else {
         iter = rows.iterator();
-        try {
-          deletePosIterator.close();
-        } catch (IOException e) {
-          throw new UncheckedIOException("Failed to close delete positions iterator", e);
-        }
       }
 
       addCloseable(iter);
+      addCloseable(deletePosIterator);
```

Comment:
Did we leak resources before?

Reply:
No, it was closed in the nested class `PositionFilterIterator` before, which is removed in this PR.